### PR TITLE
Add functionality to reset receiver link credit

### PR DESCRIPTION
**Pull request by yunhaoling**

### CODE REVIEW

File: uamqp/client.py [hunk 1]
```diff
@@ -1044,6 +1046,12 @@ def receive_message_batch(self, max_batch_size=None, on_message_received=None, t
         if len(batch) >= max_batch_size:
             return batch
 
+        if issue_link_credit_on_demand:
+            # in the case max_batch_size <= self._prefetch, no extra link credit is needed.
+            if max_batch_size > self._prefetch:
+                extra_credit_amount_needed = max_batch_size - len(batch)
+                self.message_handler.force_update_link_credit(extra_credit_amount_needed)
```
Comment (yunhaoling) [inline uamqp/client.py:1053]: separate code path for issue lssue_link_credit_on_demand case

File: uamqp/client.py [hunk 2]
```diff
@@ -1054,11 +1062,19 @@ def receive_message_batch(self, max_batch_size=None, on_message_received=None, t
...
                     expired = True
                     break
             while not self._received_messages.empty() and len(batch) < max_batch_size:
                 batch.append(self._received_messages.get())
                 self._received_messages.task_done()
+
+        if issue_link_credit_on_demand:
+            # set credit back to self._prefetch
+            # TODO: do we need to set back to the original prefetch value?
+            self.message_handler.force_update_link_credit(self._prefetch)
```
Comment (yunhaoling) [inline uamqp/client.py:1076]: no need to reset it

File: uamqp/client.py [hunk 3]
```diff
@@ -1029,10 +1029,12 @@ def receive_message_batch(self, max_batch_size=None, on_message_received=None, t
         """
         self._message_received_callback = on_message_received
         max_batch_size = max_batch_size or self._prefetch
-        if max_batch_size > self._prefetch:
-            raise ValueError(
-                'Maximum batch size cannot be greater than the '
-                'connection link credit: {}'.format(self._prefetch))
+        issue_link_credit_on_demand = kwargs.get("issue_link_credit_on_demand", False)
```
Comment (yunhaoling) [inline uamqp/client.py:1032]: renaming parameter to something like "reset_prefetch"

File: src/vendor/azure-uamqp-c/src/link.c [hunk 4]
```diff
@@ -1186,6 +1186,65 @@ int link_set_max_link_credit(LINK_HANDLE link, uint32_t max_link_credit)
...
 
+int link_reset_link_credit(LINK_HANDLE link, uint32_t link_credit, bool drain)
+{
+    int result;
+    FLOW_HANDLE flow;
+
+    if (link == NULL)
+    {
+        result = __FAILURE__;
+    }
```
Comment (annatisch) [inline src/vendor/azure-uamqp-c/src/link.c:1197]: We may need to add a check here to confirm this is a receiver link - as this operation is not legal for a sender link.

**Pull request merged**